refactor task with updated OSM inputs

thanrobinson opened the pull request.

thanrobinson committed:
refactor task with updated OSM inputs

thanrobinson commented:
![](https://github.trello.services/images/mini-trello-icon.png) [hii_road temporal automation](https://trello.com/c/LfpQnkrX/206-hiiroad-temporal-automation)

thanrobinson committed:
remove roadrun.sh, update gitignore

saanobhaai commented on src/hii_road.py:
```diff
@@ -1,358 +1,172 @@
 import argparse
 import ee
-from datetime import datetime, timezone
+from datetime import datetime, timezone, timedelta
```
No longer need `timedelta`

saanobhaai commented on src/hii_road.py:
```diff
@@ -1,358 +1,172 @@
-    scale = 300
+    scale = 100
+    OSM_START = datetime(2012, 9, 12).date()
+    KERNEL_DISTANCE = 500
```
Some comments or maybe even better some documentation in the README would help our future selves

saanobhaai commented on src/hii_road.py:
```diff
@@ -1,358 +1,172 @@
+        "water": {
             "ee_type": HIITask.IMAGE,
-            "ee_path": f"{ee_hiistatic_physical}watermask_jrc70_cciocean",
+            "ee_path": "projects/HII/v1/source/phys/watermask_jrc70_cciocean",
```
I'm not loving the `source` dir structure -- maybe once we've gone through all the drivers we can clean up.

thanrobinson replied:
added TODO to look into this

saanobhaai commented on src/hii_road.py:
```diff
@@ -1,358 +1,172 @@
+                .selfMask()
+                .multiply(self.weights["groad"])
+            )
+            .multiply(2)
```
I think we need to figure out why this is here before moving on (and either remove or add explanation)

thanrobinson replied:
I see nothing in this from either Adam's doc or the Venter paper.  The only thing I can think of, is that the weights used for OSM are either 4 or 10. In venter, the direct impact of roads is 8. Multiplying by 2 would ensure that the minimum direct influence would be 8, matching Venter.  Worth asking Adam?

saanobhaai replied:
Adam says it was an artifact of how he produced osm. I think we should remove it and generally simplify, BUT ensure that the direct and indirect weightings match Venter logic, and that the output range is "correct" with respect to that. Again here we may need to compare to other drivers and see how weightings are handled and what they mean when combined at the end.

saanobhaai commented on src/hii_road.py:
```diff
@@ -1,358 +1,172 @@
+                .selfMask()
+                .multiply(self.weights["groad"])
+            )
+            .multiply(2)
```
Any reason why these operations are outside the main chain, or can they be combined into a single set of parens like above?

thanrobinson replied:
no reason. formatting fixed to combine

saanobhaai commented on src/hii_road.py:
```diff
@@ -1,358 +1,172 @@
+            (
+                self.groads.distance(kernel=self.kernel(), skipMasked=False)
+                .lte(self.DIRECT_INFLUENCE_RADIUS)
+                .selfMask()
```
Do we need to `selfMask` and then `unmask(0)`? If so why don't we need to above? (I'm sure there are good reasons, just checking)

thanrobinson replied:
I think you are correct.  Not a necessary step. In fact, I think it does absolutely nothing in this instance.

saanobhaai commented on src/hii_road.py:
```diff
@@ -1,358 +1,172 @@
-        highway_unclassified, highway_unclassified_date = self.get_most_recent_image(
-            ee.ImageCollection(self.inputs["highway_unclassified"]["ee_path"])
+        # TODO: determine if drivers are to be normalized or not.
+        # TODO: values beyond indirect range (15km) - do we set to 0 or mask?
```
Good question -- my take is, they should be 0 because those cells have 0 influence contributing to HF coming from roads (not that they're nonexistent). But cells outside the water mask should be masked.

thanrobinson replied:
water is masked. Values beyond 15km are 0 where they fall on land.

saanobhaai commented on src/hii_road.py:
```diff
@@ -1,358 +1,172 @@
+            f"driver/roads",
         )
 
+    # Check inputs is not called if date is before OSM_START
```
Probably can lose this comment... and I don't like skipping `check_inputs` but I don't see a better way and in this particular task with these particular inputs it makes sense.

thanrobinson replied:
I agree, it seems to defeat the purpose of check_inputs.  but I also can't think of a better way at the moment.

saanobhaai reviewed (commented):
Just minor comments really -- very exciting refactor here

thanrobinson committed:
modified based on PR comments

fixed formatting, removed comments, add TODOs, removed excess .selfMask() calls

thanrobinson committed:
dynamic weighting (direct v. indirect) & (motorized v. non-motorized)

saanobhaai commented on src/hii_road.py:
```diff
@@ -3,356 +3,258 @@
+            "static": True,
         },
-        "watermask": {
+        "ocean": {
```
This is only used once, and in the same way as `water` -- shouldn't we just need one or the other, and then be consistent across tasks?

thanrobinson replied:
yes, i think we can remove 'ocean'.  it was an attempt to solve the weird error. but i don't believe it was actually part of problem

saanobhaai commented on src/hii_road.py:
```diff
@@ -3,356 +3,258 @@
+        elif impact == "INDIRECT":
+            kernel_radius = self.INDIRECT_INFLUENCE_RADIUS
+        else:
+            kernel_radius = 0
```
Personal preference but I like the pattern of putting `kernel_radius = 0` before the conditionals, and then removing the `else`

saanobhaai commented on src/hii_road.py:
```diff
@@ -3,356 +3,258 @@
-    def calc(self):
-        watermask = ee.Image(self.inputs["watermask"]["ee_path"])
+        direct_weights = road_weights.toImage()
+        indriect_weights = direct_weights.multiply(self.DIRECT_INDIRECT_INFLUENCE_RATIO)
```
sp -- `indirect_weights`

saanobhaai commented on src/hii_road.py:
```diff
@@ -3,356 +3,258 @@
 
-        self.set_aoi_from_ee('projects/HII/v1/source/realms/' + self.realm)  
+    def kernel(self, impact):
+        if impact == "DIRECT":
```
I think this way of doing it is fine. You might define these as constants up top (outside the class even):
DIRECT = "DIRECT"
INDIRECT = "INDIRECT"

thanrobinson replied:
I reformatted the kernel definition to a similar format to how i defined the SCL filters in eff_pot_hab, rather than using conditionals.

saanobhaai commented on src/hii_road.py:
```diff
@@ -3,356 +3,258 @@
+        )
 
-        roads_total = highway_total.add(roads_indirect).updateMask(watermask)
+        # # TODO: determine if drivers are to be normalized or not.
```
I think we should go ahead with this; we effectively already are. As discussed, I think if we just set another class constant up top about what the top value should be then this should be straightforward; we could make it 8 like Venter and then add weights as has been done previously, or make it 0-100 (my preference) and then in the add_weights task weight the drivers in a way that's equivalent to the relative importance assigned previously.

thanrobinson replied:
I pushed the current version. might be good to get what we have merged, and develop this logic on a new branch

thanrobinson committed:
fixed spelling, scale &, kernel definitions, removed ocean input

thanrobinson committed:
updated README

saanobhaai committed:
New version of README
